### PR TITLE
Issue 383/file upload

### PR DESCRIPTION
**Pull request by DailisLangovskis**

### CODE REVIEW

File: components/add-layers/vector/add-layers-vector.service.ts [hunk 1]
```diff
@@ -192,4 +199,49 @@ export class HsAddLayersVectorService {
       }
     }
   }
+  readUploadedFile(file: any): void {
+    let fileToJSON: any;
+    if (file.name.includes('.kml')) {
+      fileToJSON = kml(new DOMParser().parseFromString(file, 'text/xml'));
+      this.addNewLayer(fileToJSON);
+    } else if (file.name.includes('.gpx')) {
```
Comment (jmacura) [inline components/add-layers/vector/add-layers-vector.service.ts:206]:
```suggestion
    } else if (file.name.endsWith('.gpx')) {
```

File: components/add-layers/vector/add-layers-vector.service.ts [hunk 2]
```diff
@@ -192,4 +199,49 @@ export class HsAddLayersVectorService {
       }
     }
   }
+  readUploadedFile(file: any): void {
+    let fileToJSON: any;
+    if (file.name.includes('.kml')) {
```
[the remark this answers is not in the document]
Comment (DailisLangovskis) [reply components/add-layers/vector/add-layers-vector.service.ts:203]: @jmacura Yes, thank You I am well aware of this! Just made it fast, so I can Test it! Now at least I don't have to look for the answer my self! :D

File: components/add-layers/vector/add-layers-vector.service.ts [hunk 3]
```diff
@@ -192,4 +197,77 @@ export class HsAddLayersVectorService {
...
+      reader.onload = () => {
+        try {
+          const fileToJSON = JSON.parse(<string>reader.result);
+          if (fileToJSON !== undefined) {
+            if (fileToJSON.features.length > 0) {
+              this.createVectorObjectFromJson(fileToJSON);
+            }
+          }
+        } catch (e) {
+          // do nothing
```
Comment (raitisbe) [inline components/add-layers/vector/add-layers-vector.service.ts:216]: Maybe some output to console could be usefull in case of error.

**Review by raitisbe — APPROVED** [review]

**Review by jmacura — COMMENTED** [review]
The implementation is too eager. It adds the layer into the map right after the file is uploaded instead of after the click on "+ Add". So it adds the layer with some default title and once the user clicks the "+ Add", just an empty layer is created with no geometry.

**Comment by raitisbe** [comment]
> The implementation is too eager. It adds the layer into the map right after the file is uploaded instead of after the click on "+ Add". So it adds the layer with some default title and once the user clicks the "+ Add", just an empty layer is created with no geometry.

I think it's not bad that the data is visualized immediately, but to make ui behave more or less in a similar way for all the layer types, we should probably just display some overview of the uploaded data. Number of features, maybe prefill the title of show/hide the "extract styles" checkbox for kml layers. That would mean that we also need an extra button to reset the panel state when the adding is complete

**Comment by DailisLangovskis** [comment]
> > The implementation is too eager. It adds the layer into the map right after the file is uploaded instead of after the click on "+ Add". So it adds the layer with some default title and once the user clicks the "+ Add", just an empty layer is created with no geometry.
> 
> I think it's not bad that the data is visualized immediately, but to make ui behave more or less in a similar way for all the layer types, we should probably just display some overview of the uploaded data. Number of features, maybe prefill the title of show/hide the "extract styles" checkbox for kml layers. That would mean that we also need an extra button to reset the panel state when the adding is complete

In progress...

**Review by raitisbe — APPROVED** [review]

**Comment by jmacura** [comment]
Well implemented I have to say.

**Pull request merged**